hikari/impl/event_manager_base.py

```diff
@@ -50,46 +58,161 @@
-        typing.Type[event_manager.EventT_co], typing.MutableSet[WaiterT[event_manager.EventT_co]]
+    WaiterMapT = typing.Dict[typing.Type[event_manager.EventT_co], typing.Set[WaiterT[event_manager.EventT_co]]]
+
+    EventManagerBaseT = typing.TypeVar("EventManagerBaseT", bound="EventManagerBase")
+    UnboundMethodT = typing.Callable[
+        [EventManagerBaseT, gateway_shard.GatewayShard, data_binding.JSONObject],
+        typing.Coroutine[typing.Any, typing.Any, None],
     ]
 
 
+@typing.runtime_checkable
+class _FilteredMethodT(fast_protocol.FastProtocolChecking, typing.Protocol):
+    async def __call__(self, shard: gateway_shard.GatewayShard, payload: data_binding.JSONObject, /) -> None:
+        raise NotImplementedError
+
+    @property
+    def __cache_components__(self) -> config.CacheComponents:
+        raise NotImplementedError
+
+    @property
+    def __event_types__(self) -> typing.Sequence[typing.Type[base_events.Event]]:
+        raise NotImplementedError
+
+
 def _default_predicate(_: event_manager.EventT_inv) -> bool:
     return True
 
 
+def filtered(
+    event_types: typing.Union[typing.Type[base_events.Event], typing.Sequence[typing.Type[base_events.Event]]],
+    cache_components: config.CacheComponents = config.CacheComponents.NONE,
+    /,
+) -> typing.Callable[[UnboundMethodT[EventManagerBaseT]], UnboundMethodT[EventManagerBaseT]]:
+    """Add metadata to a consumer method to indicate when it should be unmarshalled and dispatched.
+
+    Parameters
+    ----------
+    event_types
+        Types of the events this raw consumer method may dispatch.
+        This may either be a singular type of a sequence of types.
+
+    Other Parameters
+    ----------------
+    cache_components : hikari.config.CacheComponents
+        Bitfield of the cache components this event may make altering calls to.
+        This defaults to `hikari.config.CacheComponents.NONE`.
+    """
+    if isinstance(event_types, typing.Sequence):
+        # dict.fromkeys is used to remove any duplicate entries here
+        event_types = tuple(dict.fromkeys(itertools.chain.from_iterable(e.dispatches() for e in event_types)))
+
+    else:
+        event_types = event_types.dispatches()
+
+    def decorator(method: UnboundMethodT[EventManagerBaseT], /) -> UnboundMethodT[EventManagerBaseT]:
+        method.__cache_components__ = cache_components  # type: ignore[attr-defined]
+        method.__event_types__ = event_types  # type: ignore[attr-defined]
+        assert isinstance(method, _FilteredMethodT), "Incorrect attribute(s) set for a filtered method"
+        return method  # type: ignore[unreachable]
+
+    return decorator
+
+
+@attr.frozen()
+class _Consumer:
+    callback: ConsumerT = attr.ib()
+    """The callback function for this consumer."""
+
+    cache_components: undefined.UndefinedOr[config.CacheComponents] = attr.ib()
+    """Bitfield of the cache components this consumer makes modifying calls to, if set."""
```

Comment:
What about storing if the cache is enabled or not in the `__init__`?

Reply:
For that _clear_enabled_ cache would have to iterate over all the registered consumers each time it's called

Reply:
> For that _clear_enabled_ cache would have to iterate over all the registered consumers each time it's called

Not exactly, it would just simplify this piece of code:
```py
        # If cache_components is NONE then it doesn't make any altering state calls.
        return (
            consumer.cache_components != config.CacheComponents.NONE
            and (consumer.cache_components & self._app.cache.settings.components) != 0
        )
```

Reply:
That code's already going to be changed multiple times in the future, already changed it locally a while back since the != NONE bit is redundant and it'll have to be changed in the future to check if self._cache before checking the cache settings

Reply:
Also are you sure the cache settings are always going to be static a cross implementations

Reply:
The way everything is setup right now, it should be. I also dont see any reason to change the cache on the fly

Reply:
> The way everything is setup right now, it should be. 

Isn't that just implementation detail of the standard implementation though